src/index.ts

```diff
@@ -0,0 +1,1368 @@
+            ? iNameToIDependList[i1Name].indexOf(i2Name) > -1
+            : i2Name === "Object";
+    }
+
+    // Some params have the type of "(DOMString or DOMString [] or Number)"
+    // we need to transform it into [“DOMString", "DOMString []", "Number"]
+    function decomposeTypes(t: string) {
+        return t.replace(/[\(\)]/g, "").split(" or ");
+    }
+
+    function createTextWriter(newLine: string) {
+        let output: string;
+        let indent: number;
+        let lineStart: boolean;
+        let stack: { content: string, indent: number }[] = [];
+
+        const indentStrings: string[] = ["", "    "];
+        function getIndentString(level: number) {
+            if (indentStrings[level] === undefined) {
+                indentStrings[level] = getIndentString(level - 1) + indentStrings[1];
+            }
+            return indentStrings[level];
+        }
+
+        function write(s: string) {
+            if (lineStart) {
+                output += getIndentString(indent);
+                lineStart = false;
+            }
+            output += s;
+        }
+
+        function reset(): void {
+            output = "";
+            indent = 0;
+            lineStart = true;
+            stack = [];
+        }
+
+        function writeLine() {
+            if (!lineStart) {
+                output += newLine;
+                lineStart = true;
+            }
+        }
+
+        reset();
+
+        return {
+
+            reset: reset,
+
+            resetIndent() { indent = 0; },
+            increaseIndent() { indent++; },
+            decreaseIndent() { indent--; },
+
+            print: write,
+            printLine(c: string) { writeLine(); write(c); },
+
+            printWithAddedIndent(c: string) { this.increaseIndent(); this.printLine(c); this.decreaseIndent(); },
+
+            clearStack() { stack = []; },
+            stackIsEmpty() { return stack.length === 0; },
+            printLineToStack(content: string) { stack.push({ content, indent }); },
+            printStackContent() {
+                stack.forEach(e => {
+                    const oldIndent = indent;
+                    indent = e.indent;
+                    this.printLine(e.content);
+                    indent = oldIndent;
+                });
+            },
+
+            getResult() { return output; }
+        };
+    }
+
+    /// Get typescript type using object dom type, object name, and it's associated interface name
+    function convertDomTypeToTsType(obj: Browser.Typed): string {
+        if (obj["override-type"]) return obj["override-type"]!;
+        if (!obj.type) throw new Error("Missing type " + JSON.stringify(obj));
+        const type = convertDomTypeToTsTypeWorker(obj);
+        return type.nullable ? makeNullable(type.name) : type.name;
+    }
+
+    function convertDomTypeToTsTypeWorker(obj: Browser.Typed): { name: string; nullable: boolean } {
+        let type;
```

Comment:
Looks like `obj["type-original"]` is never used?

This function seems to be equivalent to:
```ts
if (typeof obj.type === "string") {
    const name = covertDomTypeToTsTypeSimple(obj.type);
    const typeArg = obj.subtype && convertDomTypeToTsType(obj.subtype);
    return {
        name: typeArg ? name === "Array" ? makeArrayType(typeArg) : `${name}<${typeArg}>` : name,
        nullable: !!obj.nullable,
    };
}
else {
    assert(!obj.subtype);
    const types = obj.type.map(convertDomTypeToTsTypeWorker);
    return {
        name: types.map(t => t.name).join(" | "),
        nullable: types.some(t => t.nullable),
    };
}
```

Reply:
yes. i wanted to leave it for now to see what other idl files have

Reply:
Could still apply the above refactoring though?